x/incentives/keeper/msg_server_test.go

```diff
@@ -257,6 +257,106 @@ func (s *KeeperTestSuite) TestAddToGauge_Fee() {
+			groupFunds:           tenTokens,
+			numEpochsPaidOver:    3,
+		},
+		{
+			name:                 "user creates a perpetual group and fills group with all remaining tokens",
+			accountBalanceToFund: sdk.NewCoins(sdk.NewCoin("uosmo", osmomath.NewInt(100000000)), sdk.NewCoin(sdk.DefaultBondDenom, osmomath.NewInt(10000000))),
+			groupFunds:           tenTokens,
+			numEpochsPaidOver:    0,
+		},
+		{
+			name:                 "user creates a non-perpetual group and fills group with some remaining tokens",
+			accountBalanceToFund: sdk.NewCoins(sdk.NewCoin("uosmo", osmomath.NewInt(100000000)), sdk.NewCoin(sdk.DefaultBondDenom, osmomath.NewInt(15000000))),
+			groupFunds:           tenTokens,
+			numEpochsPaidOver:    3,
+		},
+		{
+			name:                 "module account creates a perpetual group",
+			accountBalanceToFund: sdk.Coins{},
+			groupFunds:           sdk.Coins{},
+			isModuleAccount:      true,
+			numEpochsPaidOver:    0,
+		},
+		{
+			name:                 "user tries to create a non-perpetual group but does not have enough funds to pay for the create group fee",
+			accountBalanceToFund: sdk.NewCoins(sdk.NewCoin("uosmo", osmomath.NewInt(90000000)), sdk.NewCoin(sdk.DefaultBondDenom, osmomath.NewInt(10000000))),
+			groupFunds:           tenTokens,
+			expectErr:            true,
+			numEpochsPaidOver:    3,
+		},
+		{
+			name:                 "one user tries to create a group, has enough funds to pay for the create group fee but not enough to fill the group funds",
+			accountBalanceToFund: sdk.NewCoins(sdk.NewCoin("uosmo", osmomath.NewInt(100000000)), sdk.NewCoin(sdk.DefaultBondDenom, osmomath.NewInt(9000000))),
+			groupFunds:           tenTokens,
+			expectErr:            true,
+			numEpochsPaidOver:    3,
+		},
+	}
+
+	for _, tc := range tests {
+		s.SetupTest()
+
+		testAccountPubkey := secp256k1.GenPrivKeyFromSecret([]byte("acc")).PubKey()
+		testAccountAddress := sdk.AccAddress(testAccountPubkey.Address())
+
+		ctx := s.Ctx
+		bankKeeper := s.App.BankKeeper
+		accountKeeper := s.App.AccountKeeper
+		msgServer := keeper.NewMsgServerImpl(s.App.IncentivesKeeper)
+		groupCreationFee := s.App.IncentivesKeeper.GetParams(ctx).GroupCreationFee
+
+		s.FundAcc(testAccountAddress, tc.accountBalanceToFund)
+
+		if tc.isModuleAccount {
+			testAccountAddress = accountKeeper.GetModuleAddress(types.ModuleName)
+		}
+
+		s.PrepareAllSupportedPools()
+
+		msg := &types.MsgCreateGroup{
+			Coins:             tc.groupFunds,
+			NumEpochsPaidOver: tc.numEpochsPaidOver,
+			Owner:             testAccountAddress.String(),
+			PoolIds:           []uint64{1, 2, 3},
+		}
+		// System under test.
+		_, err := msgServer.CreateGroup(sdk.WrapSDKContext(ctx), msg)
+
+		if tc.expectErr {
+			s.Require().Error(err)
+		} else {
+			s.Require().NoError(err)
+			balanceAmount := bankKeeper.GetAllBalances(ctx, testAccountAddress)
+
+			if tc.expectErr {
+				s.Require().Equal(tc.accountBalanceToFund.String(), balanceAmount.String(), "test: %v", tc.name)
```

Comment:
Is the second `tc.expectErr` reachable?

Reply:
Good catch, addressed https://github.com/osmosis-labs/osmosis/pull/6588/commits/38d1768711a7457cab229b4fd3aa4c9303ffcbc3

Removed since balance always gets reverted on error, and checking in tests does not show this fund reversion

Reply:
extracted to osmoassert